### ISSUE TITLE
Improve Tx submission: handling evictions

### ISSUE THREAD

**Issue by cmwaters** [celestiaorg/celestia-app#3768]
### Summary

The tools we provide users to assist in tx submission has come a long way and we're starting to see a more production grade system that can be relied upon by rollup sequencers or RaaS providers.

As to date, we now have:
- Local tracking of nonce to allow multiple transactions in a single block
- Multiple accounts so a single light node can submit on behalf of multiple rollups
- A TxStatus endpoint that is more lightweight and gives us more information on the lifecycle of a transaction [In progress].

### Problem Definition

There are still a few missing areas where we can improve robustness. One important one is handling evictions which cause nonce mismatches for an entire batch of transactions. We also can't rely on the consensus node we submit to for the new nonces because they could lie and cause us to break replay protection.

### Proposal

What I'm proposing is a little more elaborate. Some of the ideas I took from maelstrom which focused a lot on reliable tx submission. Essentially, the txClient needs to have a small pool (like a mempool) that tracks the transactions over their lifecycle and only discards them once they have been confirmed on chain. As transactions can potentially live forever, all transactions must have a TimeoutHeight which defaults to 5 blocks above the current height. That way as we track the blockchain we can always reach a conclusive result as to whether a transaction has failed or been successful. I propose a default of 5 that can be configured by the user.

Now, when it comes to [handling nonce mismatches](https://github.com/celestiaorg/celestia-app/blob/b684acd65178df2361c0bce6f7cab29fcc142b25/pkg/user/tx_client.go#L334), if the client knows all the unconfirmed transactions they have submitted, they can replay the transactions that they have signed to the receiving node up to the current transaction. If a transaction is evicted or times out, the other transactions can all have their nonces update accordingly (they may also need to wait past their expiration date to avoid double execution). Having the user set the expiration time also allows the user to resubmit sooner with a higher fee (although I think pricing holistically should be solved elsewhere).

**Comment by cmwaters** [comment]
There are two cases where evictions are causing the `TxClient` to submit transactions that are incorrect according to the consensus nodes local view:

1. Consensus node’s view of nonce is **greater** than `TxClient`: https://github.com/celestiaorg/celestia-app/issues/4784
2. Consensus node’s view of nonce is **less** than `TxClient`: https://github.com/celestiaorg/celestia-node/issues/3986 (used to be more common until we implemented a tracking mechanism)

For 1. this can happen if:
- the user is submitting transactions using the same keys on a different `TxClient` so it's not aware that other transactions are being committed
- The user submits a transaction. It's broadcasted, but then the consensus node evicts it (they have custom evicting logic). The node tells the `TxClient` that it is evicted and the `TxClient` rolls back their sequence by one. However afterwards, another node proposed it in a block and then the consensus node updates it's nonce to be ahead of the `TxClient`

For 2. this can happen if:
- A transaction is submitted, added to the mempool but is later removed. The `TxClient` catches if it is evicted but not if it is rejected. This can also cause all future transactions to be evicted.


### My suggested appoach:

- Handle rejections in consensus node and `TxClient`
- Only rejections result in reverting the sequence number, evictions are handled by resubmission (this shouldn't really happen because the local TTLs should be larger than the user specified one).
- Have user specified TTLs for all transactions (this should be height based to begin with but would be better time based (so the user doesn't have to query the height all the time))
    - If the users TTL expires we return to the user (they need to resubmit with a higher fee in all likelihood). This will be caught anyway as a rejection
    - We have to be wary, that one rejection could cause all future pending transactions to also be rejected. So we could add resubmit logic but because now it depends on blocks being full we may postpone that to a later point

**Comment by evan-forbes** [comment]
> Handle rejections in consensus node and TxClient

do you mean that rejections are not handled in the txClient, therefore we should handle them there or that they are also not handled in the consensus node? by handle, do you mean decrement the relevant sequences and resubmit txs?

on a side note, what is left to close the original issue?

> - Local tracking of nonce to allow multiple transactions in a single block
> - Multiple accounts so a single light node can submit on behalf of multiple rollups
> - A TxStatus endpoint that is more lightweight and gives us more information on the lifecycle of a transaction [In progress].

In general I think a minimal tx submission engine that handles sequence mismatches, timeouts, rejections, fee increases etc could be very useful for many users.

**Comment by cmwaters** [comment]
> on a side note, what is left to close the original issue?

The stuff listed is everything we currently have implemented. My proposal to close out the issues is this:

> - Handle rejections in consensus node and TxClient
> - Only rejections result in reverting the sequence number, evictions are handled by resubmission (this shouldn't really happen because the local TTLs should be larger than the user specified one).
> - Have user specified TTLs for all transactions (this should be height based to begin with but would be better time based (so the user doesn't have to query the height all the time))
    - If the users TTL expires we return to the user (they need to resubmit with a higher fee in all likelihood). This will be caught anyway as a rejection
    - We have to be wary, that one rejection could cause all future pending transactions to also be rejected. So we could add resubmit logic but because now it depends on blocks being full we may postpone that to a later point

**Comment by evan-forbes** [comment]
using user defined block height based ttls that are smaller than that of the default makes sense as an good improvement!

imo, we should increase the ttls on the consensus nodes significantly, and remove the time based ones.

**Comment by ninabarbakadze** [comment]
> using user defined block height based ttls that are smaller than that of the default makes sense as an good improvement!
> 
> imo, we should increase the ttls on the consensus nodes significantly, and remove the time based ones.

Yeah, I think we've been talking about increasing ttls on the consensus nodes for a while. We should address this as part of this issue/epic.

I might convert this into an epic and add all the smaller issues necessary in order to roll this out and for easier tracking @cmwaters

**Comment by cmwaters** [comment]
Look at tx client v2 spec which tries to resolve this. Will close this issue
